placement/algo/mirrored.go

```diff
@@ -197,16 +206,25 @@ func (a mirroredAlgorithm) ReplaceInstances(
 	leavingInstanceIDs []string,
 	addingInstances []placement.Instance,
 ) (placement.Placement, error) {
-	if err := a.IsCompatibleWith(p); err != nil {
+	err := a.IsCompatibleWith(p)
+	if err != nil {
 		return nil, err
 	}
 
 	if len(addingInstances) != len(leavingInstanceIDs) {
 		return nil, fmt.Errorf("could not replace %d instances with %d instances for mirrored replace", len(leavingInstanceIDs), len(addingInstances))
 	}
 
-	p, err := placement.MarkAllShardsAsAvailable(p)
-	if err != nil {
+	var (
+		nowNanos               = a.opts.NowFn()().UnixNano()
+		shouldCheckCutoverTime = true
+	)
+	if allLeaving(p, addingInstances, nowNanos) && allInitializing(p, leavingInstanceIDs, nowNanos) {
+		// Allow marking shards as available without checking cutover time in a reverting case.
+		shouldCheckCutoverTime = false
```

Comment:
Hmm if we are reverting a replace (i.e., removing the initializing instance and adding back the leaving instance before the shard transition happens), and we mark shards available below, wouldn't that mark the initializing shards as available and remove the leaving shards, which essentially marks the replace as complete?

Reply:
It marks the previous one as complete and then do another replace for the revert

Reply:
We do that before the shard transition happens? I thought if the shard transition hasn't happened for a replace and we perform a revert, we should just return the initializing shards and remove the leaving shards and there should be no leaving/initializing shards in the final placement. Is that not true?

Reply:
We are only reverting for add and remove right now, we could do the same for replace but for some reason we are not doing that, I can't recall why now, but adding the same reverting for replace is simple, I could do that in a follow up diff if that's what we want.

Reply:
Yeah I think that's what we'd want for replaces too. Perhaps just address it in this PR while you are there?

Reply:
Sure, sounds good